host/xtest/asym_cipher_perf.c

```diff
@@ -0,0 +1,760 @@
+
+#include "crypto_common.h"
+#include "xtest_helpers.h"
+
+static TEEC_Context ctx;
+static TEEC_Session sess;
+
+static TEEC_SharedMemory in_shm = {
+	.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT
+};
+static TEEC_SharedMemory out_shm = {
+	.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT
+};
+static TEEC_SharedMemory hash_shm = {
+	.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT
+};
+
+static void errx(const char *msg, TEEC_Result res, uint32_t *orig)
+{
+	fprintf(stderr, "%s: 0x%08x", msg, res);
+	if (orig)
+		fprintf(stderr, " (orig=%d)", (int)*orig);
+	fprintf(stderr, "\n");
+	exit (1);
+}
+
+static void check_res(TEEC_Result res, const char *errmsg, uint32_t *orig)
+{
+	if (res != TEEC_SUCCESS)
+		errx(errmsg, res, orig);
+}
+
+#define CHECK(res, name, action) do {			\
+		if ((res) != TEE_SUCCESS) {		\
+			printf(name ": 0x%08x", (res));	\
+			action				\
+		}					\
+	} while(0)
+
+static void open_ta(void)
+{
+	TEEC_UUID uuid = TA_CRYPTO_PERF_UUID;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+	uint32_t err_origin = 0;
+
+	res = TEEC_InitializeContext(NULL, &ctx);
+	check_res(res, "TEEC_InitializeContext", NULL);
+
+	res = TEEC_OpenSession(&ctx, &sess, &uuid, TEEC_LOGIN_PUBLIC, NULL,
+			       NULL, &err_origin);
+	check_res(res, "TEEC_OpenSession", &err_origin);
+}
+
+static void close_ta(void)
+{
+	TEEC_CloseSession(&sess);
+	TEEC_FinalizeContext(&ctx);
+}
+
+/*
+ * Statistics
+ *
+ * We want to compute min, max, mean and standard deviation of processing time
+ */
+
+struct statistics {
+	int n;
+	double m;
+	double M2;
+	double min;
+	double max;
+	int initialized;
+};
+
+/* Take new sample into account (Knuth/Welford algorithm) */
+static void update_stats(struct statistics *s, uint64_t t)
+{
+	double x = (double)t;
+	double delta = x - s->m;
+
+	s->n++;
+	s->m += delta / s->n;
+	s->M2 += delta * (x - s->m);
+	if (!s->initialized) {
+		s->min = s->max = x;
+		s->initialized = 1;
+	} else {
+		if (s->min > x)
+			s->min = x;
+		if (s->max < x)
+			s->max = x;
+	}
+}
+
+static double stddev(struct statistics *s)
+{
+	if (s->n < 2)
+		return NAN;
+
+	return sqrt(s->M2 / s->n);
+}
+
+static void usage(const char *progname, uint32_t width_bits, uint32_t tee_type,
+		  uint32_t mode, uint32_t salt_len, uint32_t size,
+		  uint32_t rsa_algo, int warmup, uint32_t l, uint32_t n)
+{
+	fprintf(stderr, "Usage: %s [-h]\n", progname);
+	fprintf(stderr, "Usage: %s [-t] [-m] [-k SIZE]", progname);
+	fprintf(stderr, " [-t tee_type] [-m mode] [-n LOOP] [-r|--no-inited] [-d WIDTH_BITS]");
+	fprintf(stderr, " [-k SIZE] [-a rsa_algo] [-s salt_len] [-v [-v]] [-w SEC]");
+	fprintf(stderr, "\n");
+	fprintf(stderr, "Asymmetric performance testing tool for OP-TEE\n");
+	fprintf(stderr, "\n");
+	fprintf(stderr, "Options:\n");
+	fprintf(stderr, "  -h|--help        Print this help and exit\n");
+	fprintf(stderr, "  -t TYPE          Test type for asymmetric algo [%u]\n", tee_type);
+	fprintf(stderr, "                   1:DH    2:RSA    3:ECDSA    4:ECDH    5:X25519\n");

```

Comment:
I'd rather take these parameters as strings. When this is used in a script or a saved sequence of commands it's easier to see what's tested. Same for the `-m` and `-a` parameters.

Reply:
May I keep this way to -a paramter,since there are too many cases in -a parameter.What's more, it is different for meaning of num when mode is ENC/DEC or SIGN/VERIFY

Reply:
Looping over `const char *asym_algo[] = { "DH", "RSA", "ECDSA", "ECDH", "X25519" };` to determine the index and use that can't be too hard.

Reply:
DH,RSA……are -t parameters.I mean -a parameters，such as RSAES_PKCS1_V1_5，RSAES_PKCS1_OAEP_SHA256 and so on.

Reply:
The greater reason to use the same approach there.